synapse/http/replicationagent.py

```diff
@@ -142,9 +153,10 @@ def request(
 
         This is copied from twisted.web.client.Agent, except:
 
-        * It uses a different pool key (combining the host & port).
-        * It does not call _ensureValidURI(...) since it breaks on some
-          UNIX paths.
+        * It uses a different pool key (combining the scheme with either host & port or
+          socket path).
+        * It does not call _ensureValidURI(...) since it breaks when using a worker's
+          name as a 'hostname'.
```

Comment:
> since it breaks when using a worker's name as a 'hostname'.

Is this literally when it's set as `hostname` or when someone uses are bare host?


> since it breaks on some UNIX paths.

From the before diff, is this no longer true?

Reply:
The 'some' in that statement was something my reviewer at the time suggested, based on what `_ensureValidURI()` was denying us to use. In this case, the `worker_name` took the place of a `hostname`. (See that [PR](https://github.com/matrix-org/synapse/pull/15578) for details of `worker_name` as `hostname`). The IDNA2008 standard prohibits several ascii characters that are in use by Complement for it's worker names, namely the underscore `_` and potentially the plus `+` sign, both of which work fine on the actual test setup. (See this [PR](https://github.com/matrix-org/synapse/pull/15470) for details of why IDNA2008 was being to strict, I do not believe it was detailed in the main comment though.)

Reply:
Whatever you explained isn't encompassed by the current comment very well. Perhaps something more like:

```
 * We don't call `_ensureValidURI(...)` because we use `worker_name`'s as the URL scheme and can include characters like `_` and `+` which isn't a valid strict IDNA2008 scheme but still works with all of the machinery in Synapse.
```

I mentioned scheme instead of hostname here because it seems like we're using it as the scheme according to the PR you mentioned, https://github.com/matrix-org/synapse/pull/15578/files#diff-209be8defd6cf72f8e6fd6c765ca7e11a172dc244c9f0dbcc57ece200fe53a2fR64-R65

Reply:
TLDL: Read the comment from the link as `URI = <special_scheme>+<worker_name)` not as `URI=<special_scheme(worker_name)>`.

(Long winded version that probably over explains)
Oh, that's not the scheme. The scheme comes in from the replication machinery as `synapse-replication://` and is translated into `http(s)` or `unix` a little after this spot. The actual worker's name isn't part of the scheme itself. This was just a comment to explain how this version of `request()` is different than the one from the base class(as it's an...is overload the correct word?). As this is an 'implementation' of the `Agent` class, the signature of the function isn't supposed to change.

Having said that, I feel like there is some confusion about how the URI is broken up(Twisted's doc for [URI](https://docs.twisted.org/en/stable/api/twisted.web.client.URI.html#__init__)). We are borrowing the fact that the scheme and the `netloc` (which is effectively a combination of `host:port` for normal TCP connections, and acts like a `hostname`) are definable for our own purposes. Did that help?

Reply:
I don't think it's any more clear to me. I suggest just updating the comment to be obvious and perhaps I can pick it up from there.

Reply:
>> since it breaks on some UNIX paths.

> From the before diff, is this no longer true?

I completely missed you asking about this detail. It would break on any Unix path that was outside of / as it substring the parts of that path as part of the URI path. e.g. unix://var/run/main_public.socket/_matrix/client/versions. We would look at that and intuitively know that socket would end the Unix socket path and the URI path would start after, but _ensureValidURI() would see var as the hostname/netlock and the rest as part of the URI(even with the period). It was just not handy, and since Unix sockets/schemes are outside the scope of IDNA2008 altogether...it was just another reason it had to go. But that is history now, as it's not relevant anymore. Now we use the worker_name and that need for flexibility still exists.

That probably doesn't help any more though. I guess I just don't understand what you want to know. I'm sorry.

Reply:
Maybe as changed in [e091ea2](https://github.com/matrix-org/synapse/pull/15708/commits/e091ea2df100d3248e333553636faa00d838a6e1)?

Reply:
The comment sounds plausible now :+1: